### DISCUSSION TITLE
[Show and Tell] Open Source Observer: measuring impact of OSS contributions to the OP Superchain

### DISCUSSION THREAD

**Discussion by ccerv1** [ethereum-optimism/developers#51] [Show and tell]
### Description

> Cross-posting from the [gov forum](https://gov.optimism.io/t/introducing-open-source-observer-measuring-impact-of-oss-contributions-to-the-op-superchain/7071)

With RPGF3 voting kicking off today, we are excited to present a new public good for the Optimism Community: **Open Source Observer (OSO)**

### tl;dr-

* We’ve created a [huge directory](https://github.com/opensource-observer/oss-directory) of 900+ OSS projects and nearly 20K off-/onchain artifacts
* It includes every project that applied for RPGF3 with at least a GitHub repo -- over 300!
* For every project, we've indexed a suite of metrics ... here's a [spreadsheet](https://docs.google.com/spreadsheets/d/1OEU5ExW0Tm8iEOBqXbjYLjmvhYbIHRHLhtMoNcLB2_I/edit?usp=sharing) you can start playing with
* There's even more data up on our [website](https://www.opensource.observer/project/ethereum-attestation-service) (🏗️) including project deep dives and (soon™) sortable metrics tables
* We’ll be using the data to generate voting lists for badgeholders based on quantitative impact metrics
* Jump in this [group chat](http://t.me/oso_rpgf3) if you have issues, need support, or want API access!

### Intro

OSO aims to create better visibility on the impact of different types of open source software contributions to the health of the OP Superchain.

In the long run, we want OSO to be a source of verifiable proof that “impact = profit”, at least for impact that is traceable to open source software contributions. And we want that data to be a free and easy-to-use source of ~~business~~ network intelligence for badgeholders, citizens, and builders.

### What we’ve done so far

We started this project about 3 months ago. Here are some of the highlights:

* 📖 Created a registry of projects that are contributing to Optimism
* 🖇️ Linked projects to “artifacts” such as git repos, smart contracts, and npm packages
* 💾 Indexed event data about those artifacts such as commits, contract interactions, and downloads
* 📊 Defined initial activity and impact metrics such as monthly active developers, onchain users, and package dependents
* 💡 Released an alpha version of a GraphQL API and dashboarding interface for exploring the data
* 🐶 Dogfooded the data to generate some analysis on the OSS projects applying for RPGF3 and to help power voting lists for badgeholders

### Registry

You can view the registry of 900+ OSS projects and nearly 20,000 artifacts [here](https://github.com/opensource-observer/oss-directory). Our RPGF3 coverage includes 300+ projects, 5K+ repos, 4K+ contracts, and 50 npm libraries.

The above should include all of the OSS projects in RPGF3 as well as any contribution links in their applications that specified a GitHub repo, OP Mainnet contract address, or npm package. If we missed anything (which we most certainly did), please open a PR or send us a message.

### Indexers

We built a number of indexers to capture event data from these artifacts and normalize the data. You can check them out in the [OSO monorepo](https://github.com/opensource-observer/oso). Most of the indexers are able to perform daily updates as well as historic backfills even to pre-Ethereum days. We’ll be indexing other chains and package managers soon … and are very eager to partner with other projects that have useful data to layer on. If you have ideas or notice bugs, please [open an issue](https://github.com/opensource-observer/oso/issues) or send us a message.

### Using OSO Data

Finally, we’re proud to share some different ways of accessing and composing with this data:

* 👀 Data Snapshot: we created some initial metrics for 300+ RPGF3 projects and dropped them into a [google sheet](https://docs.google.com/spreadsheets/u/0/d/1OEU5ExW0Tm8iEOBqXbjYLjmvhYbIHRHLhtMoNcLB2_I/edit). Have fun with it!
* 🎉 API: We have a GraphQL API that you can use to make queries directly. Send us a message if you want access – we don’t have much in the way of docs but are happy to provide as much support as we can manage at this time.
* 📈 Dashboard: We have a dashboard interface running on top of the API available at [opensource.observer](https://www.opensource.observer/). You can analyze cohorts of projects or deep dive on specific project metrics. We released this ahead of schedule, but it’s still got some performance issues and obvious bugs we’re squashing.
* ✨ Lists: We have an [insights repo](https://github.com/opensource-observer/insights) for generating voting lists. We’ve populated it with some initial strategies, like ranking projects based on the number of [stars per active contributor](https://github.com/opensource-observer/insights/blob/main/notebooks/2023-11-04_StarsPerContributor.ipynb) or the [time delta between](https://github.com/opensource-observer/insights/blob/main/notebooks/2023-11-04_StarsPerContributor.ipynb) the project’s first commit and its first transaction on OP Mainnet. We’ll be patching gaps in the data and refining these lists over the next week.

### Sustainability

Ray, Reuven and I began this work while employed at Protocol Labs working on [hypercerts](https://hypercerts.org). We left in October to work on this full-time. Our new company is called Kariba Labs. OSO is the first thing we're releasing. 

We are taking an "Open Source, Open Data, Open Infra" approach to what we build. That means all our code is developed using permissive licenses (e.g. MIT/Apache 2.0). All collected and processed data will be openly shared with the community (to the extent allowed by terms of service). And we will open up our infrastructure for anyone to contribute or build upon our existing infrastructure at cost.

In terms of funding, we received a [Builders Grant](https://app.charmverse.io/op-grants/page-37085606550601957) in July and won the “Best App for RetroPGF, Public Goods, or Measuring Impact” track at [Superhack](https://ethglobal.com/events/superhack/prizes/optimism-8x4eb) in August for an early implementation of OSS Directory. We also secured some seed funding from the Protocol Labs Network.

### Community

If you want to connect, discuss an issue, or follow updates, join our Telegram chat [here](http://t.me/oso_rpgf3). We just created it, so there’s no one in it at the time of writing 😂 but we'd love to see you there!

**Comment by sbvegan** [comment]
This is such an awesome project, thank you so much for sharing!

**Comment by opfocus** [comment]
Great, I'm wondering how it can benefit other developers besides tracking project metrics. Without a doubt, I believe it has vast potential for expansion.

**Reply by ccerv1** [reply]
Check out some initial thoughts on this in our first blog post [here](https://kariba.substack.com/p/open-source-open-data-open-infra)

**Reply by opfocus** [reply]
A great reading experience, I think I will read it at a different time. I wish your projects greater success
